File: expandablrecyclerview/src/main/java/com/umut/expandablrecyclerview/adapter/index/CachedExpandableIndexProvider.java
package com.umut.expandablrecyclerview.adapter.index;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import android.util.SparseArray;
import android.util.SparseIntArray;

import com.umut.expandablrecyclerview.adapter.ChildCoordinate;

public class CachedExpandableIndexProvider implements ExpandableIndexProvider {

    protected static final int UNSPECIFIED = -1;

    private int totalItemCount;
    @NonNull
    private final SparseIntArray parentPositionCache;
    @NonNull
    private final SparseArray<ChildCoordinate> childCoordinateCache;
    @NonNull
    private final SparseIntArray viewTypeCache;

    public CachedExpandableIndexProvider() {
        this.parentPositionCache = new SparseIntArray();
        this.childCoordinateCache = new SparseArray<>();
        this.viewTypeCache = new SparseIntArray();
        totalItemCount = UNSPECIFIED;
    }

    @Override
    public int getTotalItemCount() {
        return totalItemCount;
    }

    @Override
    public int getAdapterIndexForParentAt(int adapterParentIndex) {
        throw new UnsupportedOperationException("This operation is not supported.");
    }

    @Override
    public int getViewType(int position) {
        return viewTypeCache.get(position, UNSPECIFIED);
    }

    @Override
    public int parentIndexFromAdapterPosition(int adapterPosition) {
        return parentPositionCache.get(adapterPosition, UNSPECIFIED);
    }

    @Override
    public boolean isExpanded(int dataPosition) {
        throw new UnsupportedOperationException("This operation is not supported");
    }

    @Override
    public void onExpand(int dataPosition) {
        throw new UnsupportedOperationException("This operation is not supported");
    }

    @Override
    public void onCollapse(int dataPosition) {
        throw new UnsupportedOperationException("This operation is not supported");
    }

    @Override
    public ChildCoordinate getChildCoordinateFromAdapterIndex(int adapterPosition) {
        return childCoordinateCache.get(adapterPosition);
    }

    void flushCache() {
        totalItemCount = UNSPECIFIED;
        parentPositionCache.clear();
        childCoordinateCache.clear();
        viewTypeCache.clear();
    }

    void cacheParentPosition(int position, int parentIndex) {
        parentPositionCache.append(position, parentIndex);
    }

    void cacheChildCoordinate(int position, @Nullable ChildCoordinate coordinate) {
        childCoordinateCache.append(position, coordinate);
    }

    void cacheTotalItemCount(int totalItemCount) {
        this.totalItemCount = totalItemCount;
    }

    void cacheViewType(int position, @ViewType int viewType) {
        this.viewTypeCache.append(position, viewType);
    }
}

File: expandablrecyclerview/src/main/java/com/umut/expandablrecyclerview/adapter/index/CachedExpandableIndexProvider.kt
package com.umut.expandablrecyclerview.adapter.index

import android.util.SparseArray
import android.util.SparseIntArray

import com.umut.expandablrecyclerview.adapter.ChildCoordinate
import com.umut.expandablrecyclerview.adapter.index.ExpandableIndexProvider.Companion.UNSPECIFIED

class CachedExpandableIndexProvider : ExpandableIndexProvider {

    private var totalItemCount: Int = 0
    private val parentPositionCache: SparseIntArray = SparseIntArray()
    private val childCoordinateCache: SparseArray<ChildCoordinate> = SparseArray()
    private val viewTypeCache: SparseIntArray = SparseIntArray()

    init {
        totalItemCount = UNSPECIFIED
    }

    override fun getTotalItemCount(): Int {
        return totalItemCount
    }

    override fun getAdapterIndexForParentAt(adapterParentIndex: Int): Int {
        throw UnsupportedOperationException("This operation is not supported.")
    }

    override fun getViewType(position: Int): Int {
        return viewTypeCache.get(position, UNSPECIFIED)
    }

    override fun parentIndexFromAdapterPosition(adapterPosition: Int): Int {
        return parentPositionCache.get(adapterPosition, UNSPECIFIED)
    }

    override fun isExpanded(dataPosition: Int): Boolean {
        throw UnsupportedOperationException("This operation is not supported")
    }

    override fun onExpand(dataPosition: Int) {
        throw UnsupportedOperationException("This operation is not supported")
    }

    override fun onCollapse(dataPosition: Int) {
        throw UnsupportedOperationException("This operation is not supported")
    }

    override fun getChildCoordinateFromAdapterIndex(adapterPosition: Int): ChildCoordinate? {
        return childCoordinateCache.get(adapterPosition)
    }

    internal fun flushCache() {
        totalItemCount = UNSPECIFIED
        parentPositionCache.clear()
        childCoordinateCache.clear()
        viewTypeCache.clear()
    }

    internal fun cacheParentPosition(position: Int, parentIndex: Int) {
        parentPositionCache.append(position, parentIndex)
    }

    internal fun cacheChildCoordinate(position: Int, coordinate: ChildCoordinate?) {
        childCoordinateCache.append(position, coordinate)
    }

    internal fun cacheTotalItemCount(totalItemCount: Int) {
        this.totalItemCount = totalItemCount
    }

    internal fun cacheViewType(position: Int, @ExpandableIndexProvider.ViewType viewType: Int) {
        this.viewTypeCache.append(position, viewType)
    }
}
